plutus-contract/src/Plutus/Contract/Test/DynamicLogic/Quantify.hs

```diff
@@ -0,0 +1,192 @@
+
+-- | A `Quantification` over a type @a@ is a generator that can be used with
+--   `Plutus.Contract.Test.ContractModel.forAllQ` to generate random values in
+--   DL scenarios. In addition to a QuickCheck generator a `Quantification` contains a shrinking
+--   strategy that ensures that shrunk values stay in the range of the generator.
+data Quantification a = Quantification
+  { genQ :: Maybe (Gen a),
+    isaQ :: a -> Bool,
+    shrQ :: a -> [a]
+  }
+
+isEmptyQ :: Quantification a -> Bool
+isEmptyQ = not . isJust . genQ
+
+generateQ :: Quantification a -> Gen a
+generateQ q = fromJust (genQ q) `suchThat` isaQ q
+
+shrinkQ :: Quantification a -> a -> [a]
+shrinkQ q a = filter (isaQ q) (shrQ q a)
+
+-- | Pack up an `Arbitrary` instance as a `Quantification`. Treats all values as being in range.
+arbitraryQ :: Arbitrary a => Quantification a
+arbitraryQ = Quantification (Just arbitrary) (const True) shrink
+
+-- | Generates exactly the given value. Does not shrink.
+exactlyQ :: Eq a => a -> Quantification a
+exactlyQ a = Quantification
+    (Just $ return a)
+    (==a)
+    (const [])
+
+-- | Generate a random value in a given range (inclusive).
+chooseQ :: (Arbitrary a, Random a, Ord a) => (a, a) -> Quantification a
+chooseQ r@(a, b) = Quantification
+    (guard (a <= b) >> (Just $ choose r))
+    is
+    (filter is . shrink)
+    where is x = a <= x && x <= b
+
+-- | Pick a random value from a list. Treated as an empty choice if the list is empty:
+--
+-- @
+-- `Plutus.Contract.Test.ContractModel.forAllQ` (`elementsQ` []) == `Control.Applicative.empty`
+-- @
+elementsQ :: Eq a => [a] -> Quantification a
+elementsQ as = Quantification g (`elem` as) (\a -> takeWhile (/=a) as)
+    where g | null as   = Nothing
+            | otherwise = Just (elements as)
+
+-- | Choose from a weighted list of quantifications. Treated as an `Control.Applicative.empty`
+--   choice if no quantification has weight > 0.
+frequencyQ :: [(Int, Quantification a)] -> Quantification a
+frequencyQ iqs =
+    Quantification
+        (case [(i, g) | (i, q) <- iqs, i > 0, Just g <- [genQ q]] of
+            []  -> Nothing
+            igs -> Just (frequency igs))
+        (isa iqs)
+        (shr iqs)
+    where isa [] _           = False
+          isa ((i, q):iqs) a = (i > 0 && isaQ q a) || isa iqs a
+          shr [] _ = []
+          shr ((i, q):iqs) a = [a' | i > 0, isaQ q a, a' <- shrQ q a]
+                               ++ shr iqs a
+
+-- | Choose from a list of quantifications. Same as `frequencyQ` with all weights the same (and >
+--   0).
+oneofQ :: [Quantification a] -> Quantification a
+oneofQ qs = frequencyQ $ map (1, ) qs
+
+-- | `Quantification` is not a `Functor`, since it also keeps track of the range of the generators.
+--   However, if you have two functions
+-- @
+-- to   :: a -> b
+-- from :: b -> a
+-- @
+--   satisfying @from . to = id@ you can go from a quantification over @a@ to one over @b@. Note
```

Comment:
Oof, that's quite the condition. Maybe it should take a `Control.Lens.Iso`.

Reply:
It's not quite an iso though, you only need an embedding of `a` into `b`.